File: Elixir/1_Two_Sum.ex
defmodule Solution do
  @spec two_sum(nums :: [integer], target :: integer) :: [integer]
  def two_sum(nums, target) do
    two_sum_helper(nums, target, 0, %{}, [])
  end

  defp two_sum_helper([], _, _, _, _), do: []  # Base case: Empty list, no solution found
  defp two_sum_helper([num | rest], target, index, seen, result) do
    complement = target - num
    case Map.get(seen, complement) do
      nil -> 
        new_seen = Map.put(seen, num, index)
        two_sum_helper(rest, target, index + 1, new_seen, result)
      other -> [other, index]
    end
  end
end

File: Erlang/1_Two_Sum.erl
-spec two_sum([integer()], integer()) -> [integer()].
two_sum(Nums, Target) ->
    two_sum_helper(Nums, Target, 0, #{}, []).

two_sum_helper([], _, _, _, _) ->
    []; % Base case: Empty list, no solution found
two_sum_helper([Num | Rest], Target, Index, Seen, Result) ->
    Complement = Target - Num,
    case maps:is_key(Complement, Seen) of
        true ->
            [maps:get(Complement, Seen), Index];
        false ->
            NewSeen = maps:put(Num, Index, Seen),
            two_sum_helper(Rest, Target, Index + 1, NewSeen, Result)
    end.
